src/OpticalFlowEstimator.cpp

```diff
@@ -0,0 +1,598 @@
+    for (size_t i = 0; i < tails.size(); ++i) {
+        if (status[i]
+         && tails[i].x > image_size.width  * x_cutoff
+         && tails[i].x < image_size.width  * (1.0 - x_cutoff)
+         && tails[i].y > image_size.height * y_cutoff
+         && tails[i].y < image_size.height * (1.0 - y_cutoff)) {
+
+            cv::Point p = tails[i];
+            cv::Point q = heads[i];
+            total_x += p.x - q.x;
+            total_y += p.y - q.y;
+            num_points++;
+        }
+    }
+
+    cv::Point2f avg_vector;
+    avg_vector.x = total_x / static_cast<double>(num_points);
+    avg_vector.y = total_y / static_cast<double>(num_points);
+
+    return avg_vector;
+}
+
+void OpticalFlowEstimator::findFeatureVectors(
+        const cv::gpu::GpuMat& curr_frame,
+        const cv::gpu::GpuMat& /*curr_gray_frame*/,
+        const cv::gpu::GpuMat& /*last_frame*/,
+        const cv::gpu::GpuMat& last_gray_frame,
+        std::vector<cv::Point2f>& tails,
+        std::vector<cv::Point2f>& heads,
+        std::vector<uchar>& status,
+        bool debug) const
+{
+    int64 start = cv::getTickCount();
+
+    // Create the feature detector and perform feature detection
+    cv::gpu::GoodFeaturesToTrackDetector_GPU detector(
+                    flow_estimator_settings_.points,
+                    flow_estimator_settings_.quality_level,
+                    flow_estimator_settings_.min_dist);
+    cv::gpu::GpuMat d_prev_pts;
+
+    detector(last_gray_frame, d_prev_pts);
+
+    if (debug_settings_.debug_times) {
+        ROS_WARN("post detector: %f",
+                 (cv::getTickCount() - start) / cv::getTickFrequency());
+    }
+
+    // Create optical flow object
+    cv::gpu::PyrLKOpticalFlow d_pyrLK;
+
+    d_pyrLK.winSize.width = flow_estimator_settings_.win_size;
+    d_pyrLK.winSize.height = flow_estimator_settings_.win_size;
+    d_pyrLK.maxLevel = flow_estimator_settings_.max_level;
+    d_pyrLK.iters = flow_estimator_settings_.iters;
+
+    cv::gpu::GpuMat d_next_pts;
+    cv::gpu::GpuMat d_status;
+
+    d_pyrLK.sparse(last_scaled_image_,
+                   curr_frame,
+                   d_prev_pts,
+                   d_next_pts,
+                   d_status);
+
+    if (debug_settings_.debug_times) {
+        ROS_WARN("PYRLK sparse: %f",
+                 (cv::getTickCount() - start) / cv::getTickFrequency());
+    }
+
+    tails.resize(d_prev_pts.cols);
+    download(d_prev_pts, tails);
+
+    heads.resize(d_next_pts.cols);
+    download(d_next_pts, heads);
+
+    status.resize(d_status.cols);
+    download(d_status, status);
+
+    // Publish debugging image with all vectors drawn
+    if (debug && debug_settings_.debug_vectors_image) {
+        // Draw arrows
+        cv::Mat arrow_image;
+        curr_frame.download(arrow_image);
+        drawArrows(arrow_image,
+                   tails,
+                   heads,
+                   status,
+                   cv::Scalar(255, 0, 0));
```

Comment:
You set the default value in a in the function definition. Can we use that?

Reply:
Fixed